completions_test.go

```diff
@@ -2691,3 +2691,171 @@ func TestFixedCompletions(t *testing.T) {
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Test that when all flags in a group are present, flags are not suggested without the - prefix
+	output, err = executeCommand(rootCmd, ShellCompNoDescRequestCmd, "child",
+		"--group1-1", "8",
+		"--group1-2", "value2",
+		"--group1-3",
+		"")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	expected = strings.Join([]string{
+		"subArg",
+		":4",
+		"Completion ended with directive: ShellCompDirectiveNoFileComp", ""}, "\n")
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+}
+
+func TestCompletionForMutuallyExclusiveFlags(t *testing.T) {
+	rootCmd := &Command{
+		Use: "root",
+		Run: emptyRun,
+	}
+	childCmd := &Command{
+		Use: "child",
+		ValidArgsFunction: func(cmd *Command, args []string, toComplete string) ([]string, ShellCompDirective) {
+			return []string{"subArg"}, ShellCompDirectiveNoFileComp
+		},
+		Run: emptyRun,
+	}
+	rootCmd.AddCommand(childCmd)
+
+	rootCmd.PersistentFlags().IntSlice("group1-1", []int{1}, "group1-1")
+	rootCmd.PersistentFlags().String("group1-2", "", "group1-2")
+
+	childCmd.Flags().Bool("group1-3", false, "group1-3")
+	childCmd.Flags().Bool("flag2", false, "flag2")
+
+	// Add flags to a group
+	childCmd.MarkFlagsMutuallyExclusive("group1-1", "group1-2", "group1-3")
+
+	// Test that flags in a mutually exclusive group are not suggested without the - prefix
+	output, err := executeCommand(rootCmd, ShellCompNoDescRequestCmd, "child", "")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	expected := strings.Join([]string{
+		"subArg",
+		":4",
+		"Completion ended with directive: ShellCompDirectiveNoFileComp", ""}, "\n")
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Test that flags in a mutually exclusive group are suggested with the - prefix
+	output, err = executeCommand(rootCmd, ShellCompNoDescRequestCmd, "child", "-")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	expected = strings.Join([]string{
+		"--group1-1",
+		"--group1-2",
+		"--flag2",
+		"--group1-3",
+		":4",
+		"Completion ended with directive: ShellCompDirectiveNoFileComp", ""}, "\n")
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Test that when a flag in a mutually exclusive group is present, the other flags in the group are
+	// not suggested even with the - prefix
+	output, err = executeCommand(rootCmd, ShellCompNoDescRequestCmd, "child", "--group1-1", "8", "-")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	expected = strings.Join([]string{
+		"--group1-1", // Should be repeated since it is a slice
+		"--flag2",
+		":4",
+		"Completion ended with directive: ShellCompDirectiveNoFileComp", ""}, "\n")
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+}
```

Comment:
We should probably include a test case for the persistent/local mixing and ensure it behaves on both the parent and subcommand.

On the parent (which is missing the subcommands local flags) the flag groups shouldn't apply at all.

On the subcommand things should behave as normal.

Reply:
Will do